Api provides admin functionality to create article

LucasKnudsen opened the pull request:
[**PT-Story:** ](https://www.pivotaltracker.com/story/show/178092075) 
### User Story 
``` 
As an Api
In order for an journalist to create an article
I need to be able to handle a submitted form and create an article based on this data 
``` 
## Changes proposed in this pull request: 
* Adds create action for Article controller
* Limits article creation to 'journalist' role
* Adds error handlers for unauthenticated user, for unauthorized user, and for lacking params.
## What I have learned working on this feature: 
* Mass assignment with .permit()
* That render json: doesn't halt execution, but we can use `and return` for this instead
* render json: in private methods will still continue on over to the main action method
* Role based conditionals

LucasKnudsen committed:
Writes request spec for article creation

LucasKnudsen committed:
wraps params in article, formats controller, adds role_authenticator functionality

christerforsgren91 committed:
Add method to create an article

christerforsgren91 committed:
Add error message if a member try to create an article

christerforsgren91 committed:
Adds cant create article if your not authenticated also if your not a journalist

arunbhalli committed:
Add if params are missing then return error message

LucasKnudsen committed:
Adds category to permitted params in controller

tochman commented on app/controllers/api/articles_controller.rb:
```diff
@@ -17,9 +19,21 @@ def index
   end
+
+  def create
+    unless current_user.role == "journalist"
```
If you are using `enum` you can simplify this

```ruby
current_user.journalist?
```

LucasKnudsen replied:
Good point!

tochman commented on app/controllers/api/articles_controller.rb:
```diff
@@ -17,9 +19,21 @@ def index
+      render json: { error_message: "You dont have access" }, status: 403 and return
+    end
+
+    article = current_user.articles.create(params[:article].permit(:title, :teaser, :body, :category))
```
Not a blocker, but a good practice is to extract this to a `private` method

```ruby
def article_params
   params[:article].permit(:title, :teaser, :body, :category)
end
```

and then use it:

```ruby
article = current_user.articles.create(article_params)
```

LucasKnudsen replied:
Good point x2!

tochman commented on spec/requests/api/auth/journalist_can_create_articles_spec.rb:
```diff
@@ -0,0 +1,110 @@
+               teaser: "Some damn teaser",
+               body: "Husband found dead allegedly because he wasn't testing first",
+               category: "Hollywood",
+               user_id: journalist.id,
```
There is no need to pass in `user_id`, you are authenticating with a token and this attribute is not used in your implementation.

LucasKnudsen replied:
Ye, not sure what's going on there actually :3

tochman commented on spec/requests/api/auth/journalist_can_create_articles_spec.rb:
```diff
@@ -0,0 +1,110 @@
+    end
+
+    it "is expected that member cant create article" do
+      expect(response_json["error_message"]).to eq "You dont have access"
```
`dont`. -> `don't` 

But the whole message feels cryptic to me

lukeperrin10 commented on spec/requests/api/auth/journalist_can_create_articles_spec.rb:
```diff
@@ -0,0 +1,110 @@
+    end
+  end
+
+  describe "unccessfully" do
```
Just a small typo here on line 65

lukeperrin10 commented on spec/requests/api/auth/journalist_can_create_articles_spec.rb:
```diff
@@ -0,0 +1,110 @@
+      expect(response_json["errors"].first).to eq "You need to sign in or sign up before continuing."
+    end
+  end
+  describe "unccessfully with missing params" do
```
Here aswell

LucasKnudsen replied:
we unccessed :(

LucasKnudsen committed:
Refactors enum condition,
removes user_id from tests as it is not used,
rewrites error message to something less cryptic,
refactors article creation params into a private method

LucasKnudsen committed:
Fixes spelling mistakes

1CIM reviewed (commented):
looks like a little different solar graph settings.

tochman reviewed (approved):
Looks great 🤩
